### PR TITLE
[rllib] Refactor Multi-GPU for PPO

### PR DESCRIPTION
**Pull request by victorsun123**
## What do these changes do?
Refactor Multi-GPU support for PPO algorithm

Main notes:
The each evaluator, including the local evaluator has a copy of the graph.
The MultiGPU Optimizer has also a copy of the graph, with `N` gradient op copies, where N is the number of devices exposed.

Before and after each optimizer step, the many copies of the graph are synchronized. 

## TODOS:
 - [x]  `python /ray/python/ray/rllib/train.py --env CartPole-v1 --run PPO --stop '{"training_iteration": 2}' --config '{"kl_coeff": 1.0, "num_sgd_iter": 10, "sgd_stepsize": 1e-4, "sgd_batchsize": 64, "timesteps_per_batch": 2000, "num_workers": 1, "use_gae": false}'` is failing

 - [ ] Comments/cleanup
 - [x] YAPF + Lint
 - [ ] With matching weights and dataset, the graph outputs the same loss and KL and entropy. However, the resulting weights are different after 1 SGD step (and with changing the optimizer with SGD instead of Adam). 
     - Setting extra_ops vs not setting extra_ops changes the gradient calculation?

### CODE REVIEW

**Comment by ericl** [comment]
Looks like this needs to be rebased (also some PAL stuff leaked in).

File: python/ray/rllib/dqn/dqn_replay_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
+from __future__ import absolute_import
```
[the remark this answers is not in the document]
Comment (robertnishihara) [reply python/ray/rllib/dqn/dqn_replay_evaluator.py]: The file? Or the line `from __future__ import absolute_import`? We include that line at the top of every Python file.

[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/rllib/dqn/dqn_replay_evaluator.py]: ah this comment is for the file

**Comment by robertnishihara** [comment]
@victorsun123 is this still in progress?

**Comment by richardliaw** [comment]
I have to clean it up - yeah

**Comment by michaeltu1** [comment]
GPU ids are not detected

File: python/ray/rllib/ppo/loss.py [hunk 2]
```diff
@@ -90,3 +91,19 @@ def compute(self, observation):
...
+    def extra_ops(self):
+        # TODO(rliaw): Note - need to reweight these means
+        return {
+            "loss": self.loss,
+            "policy_loss": self.mean_policy_loss,
+            "vf_loss": self.mean_vf_loss,
+            "kl": self.mean_kl,
+            "entropy": self.mean_entropy}
+
+    def update_kl(self, new_kl):
```
Comment (richardliaw) [inline python/ray/rllib/ppo/loss.py]: this is a duplicate I think

Comment (michaeltu1) [reply python/ray/rllib/ppo/loss.py]: agreed

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 3]
```diff
@@ -7,11 +7,15 @@
...
 
 import ray
+
 from ray.rllib.optimizers.policy_evaluator import TFMultiGPUSupport
 from ray.rllib.optimizers.policy_optimizer import PolicyOptimizer
 from ray.rllib.optimizers.sample_batch import SampleBatch
 from ray.rllib.optimizers.multi_gpu_impl import LocalSyncParallelOptimizer
 from ray.rllib.utils.timer import TimerStat
+from ray.rllib.utils.filter import MeanStdFilter
+from collections import OrderedDict
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/multi_gpu.py]: should move this up to non-ray imports

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 4]
```diff
@@ -26,12 +30,15 @@ class LocalMultiGPUOptimizer(PolicyOptimizer):
...
-    def _init(self, sgd_batch_size=128, sgd_stepsize=5e-5, num_sgd_iter=10):
+    def _init(self, sgd_batch_size=128, sgd_stepsize=5e-5,
+              num_sgd_iter=10, init_extra_ops=False):
         assert isinstance(self.local_evaluator, TFMultiGPUSupport)
         self.batch_size = sgd_batch_size
         self.sgd_stepsize = sgd_stepsize
         self.num_sgd_iter = num_sgd_iter
         gpu_ids = ray.get_gpu_ids()
+        # TODO(rliaw): fix this hack!
+        gpu_ids = os.environ["CUDA_VISIBLE_DEVICES"].split(",")
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/multi_gpu.py]: Remove this but we need to make sure GPUs are utilized for both py2.7 and py3.6 on the newest ray

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 5]
```diff
@@ -60,51 +67,97 @@ def _init(self, sgd_batch_size=128, sgd_stepsize=5e-5, num_sgd_iter=10):
...
             assert isinstance(samples, SampleBatch)
 
+        self.num_steps_sampled += samples.count
+        self.num_steps_trained += samples.count
+
+        if postprocess_samples:
+            samples = postprocess_samples(samples)
+
         with self.load_timer:
+            # TODO(rliaw): introduce support for printing out trace
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/multi_gpu.py]: renive

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 6]
```diff
@@ -60,51 +67,97 @@ def _init(self, sgd_batch_size=128, sgd_stepsize=5e-5, num_sgd_iter=10):
...
             tuples_per_device = self.par_opt.load_data(
                 self.local_evaluator.sess,
                 samples.columns([key for key, _ in self.loss_inputs]))
 
+        names = [
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/multi_gpu.py]: Let's just get rid of this (at the end; it helps for debugging right now)

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 7]
```diff
@@ -60,51 +67,97 @@ def _init(self, sgd_batch_size=128, sgd_stepsize=5e-5, num_sgd_iter=10):
...
+                        stat(output)
+                        self.loss_stats[k] = np.mean(stat.rs.mean)
+
+                loss = self.loss_stats["loss"]
+                policy_loss = self.loss_stats["policy_loss"]
+                vf_loss = self.loss_stats["vf_loss"]
+                kl = self.loss_stats["kl"]
+                entropy = self.loss_stats["entropy"]
+
+                print(
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/multi_gpu.py]: Let's just get rid of this (at the end; it helps for debugging right now)

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 8]
```diff
@@ -60,51 +67,97 @@ def _init(self, sgd_batch_size=128, sgd_stepsize=5e-5, num_sgd_iter=10):
...
 
-        self.num_steps_sampled += samples.count
-        self.num_steps_trained += samples.count
+                    outputs = self.par_opt.optimize(
+                        self.sess,
+                        permutation[batch_index] * self.per_device_batch_size,
+                        extra_ops=self.extra_ops,
+                        extra_feed_dict=self.local_evaluator.extra_feed_dict())
+                    for output, k, stat in zip(outputs,
+                                               self.extra_stats.keys(),
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/multi_gpu.py]: nit: `for output, (k, stat) in zip(outputs, self.extra_stats.items()):`

File: python/ray/rllib/optimizers/multi_gpu_impl.py [hunk 9]
```diff
@@ -7,6 +7,8 @@
 
 from tensorflow.python.client import timeline
 import tensorflow as tf
+from tensorflow.python import debug as tf_debug
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/multi_gpu_impl.py]: remove

File: python/ray/rllib/optimizers/multi_gpu_impl.py [hunk 10]
```diff
@@ -168,6 +175,7 @@ def optimize(self, sess, batch_index, extra_ops=[], extra_feed_dict={},
 
         feed_dict = {self._batch_index: batch_index}
         feed_dict.update(extra_feed_dict)
+        #sess = tf_debug.LocalCLIDebugWrapperSession(sess)
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/multi_gpu_impl.py]: remove

File: python/ray/rllib/ppo/loss.py [hunk 11]
```diff
@@ -12,6 +13,7 @@ class ProximalPolicyGraph(object):
     other_output = ["vf_preds", "logprobs"]
     is_recurrent = False
 
+    # TODO(rliaw): clean up this initializer
```
Comment (richardliaw) [inline python/ray/rllib/ppo/loss.py]: remove

File: python/ray/rllib/ppo/loss.py [hunk 12]
```diff
@@ -93,3 +94,12 @@ def get_initial_state(self):
 
     def loss(self):
         return self.loss
+
+    def extra_ops(self):
+        # TODO(rliaw): Note - need to reweight these means
```
Comment (richardliaw) [inline python/ray/rllib/ppo/loss.py]: remove this TODO

File: python/ray/rllib/ppo/ppo.py [hunk 13]
```diff
@@ -46,7 +45,7 @@
         "inter_op_parallelism_threads": 2,
     },
     # Batch size for policy evaluations for rollouts
-    "rollout_batchsize": 1,
+    "rollout_batchsize": 1,  # TODO(rliaw): why is this so low
```
Comment (richardliaw) [inline python/ray/rllib/ppo/ppo.py]: Can you look into this?

File: python/ray/rllib/ppo/ppo.py [hunk 14]
```diff
@@ -55,7 +54,7 @@
     "entropy_coeff": 0.0,
     # PPO clip parameter
     "clip_param": 0.3,
-    # Target value for KL divergence
+    # Target value for = divergence
```
Comment (richardliaw) [inline python/ray/rllib/ppo/ppo.py]: revert

File: python/ray/rllib/ppo/ppo.py [hunk 15]
```diff
@@ -281,7 +187,7 @@ def _save(self, checkpoint_dir):
         extra_data = [
             self.local_evaluator.save(),
             self.global_step,
-            self.kl_coeff,
+            self.kl_coeff,   # TODO(rliaw): move this to local evaluator
```
Comment (richardliaw) [inline python/ray/rllib/ppo/ppo.py]: can you find where kl_coeff is being handled? It should be updated, but not sure where the final is kept... It may be either in PPO.py, local_evaluator, or nowhere.

File: python/ray/rllib/utils/sampler.py [hunk 16]
```diff
@@ -4,6 +4,8 @@
 
 from collections import namedtuple
 import numpy as np
+import ray
```
Comment (richardliaw) [inline python/ray/rllib/utils/sampler.py]: ray imports should be below other imports, separated by a space

File: python/ray/rllib/train.py [hunk 17]
```diff
@@ -57,6 +59,8 @@
 
 
 if __name__ == "__main__":
+    np.random.seed(0)
```
Comment (richardliaw) [inline python/ray/rllib/train.py]: remove

**Review by richardliaw — COMMENTED** [review]
Left some general comments. I would run `ray/.travis/yapf.sh` on `rllib/` and fix those changes accordingly too.

File: python/ray/rllib/ppo/ppo.py (hunk 15 above)
Comment (michaeltu1) [reply python/ray/rllib/ppo/ppo.py]: ppo.py L143: "info["kl_coefficient"] = self.local_evaluator.update_kl(info['kl'])"
should I change it to "info["kl"] = self.local_evaluator.update_kl(info['kl'])" instead?

Comment (richardliaw) [reply python/ray/rllib/ppo/ppo.py]: This is a different issue right?

Comment (richardliaw) [reply python/ray/rllib/ppo/ppo.py]: Let's remove `kl_coeff` in this class

File: python/ray/rllib/optimizers/policy_evaluator.py [hunk 18]
```diff
@@ -145,3 +145,6 @@ def build_tf_loss(self, input_placeholders):
         """
 
         raise NotImplementedError
+
+    def tf_extra_ops(self):
+        return []
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/policy_evaluator.py]: this should be a dict rather than a list I think

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 19]
```diff
@@ -26,7 +29,8 @@ class LocalMultiGPUOptimizer(PolicyOptimizer):
     the TFMultiGPUSupport API.
     """
 
-    def _init(self, sgd_batch_size=128, sgd_stepsize=5e-5, num_sgd_iter=10):
+    def _init(self, sgd_batch_size=128, sgd_stepsize=5e-5,
+              num_sgd_iter=10, init_extra_ops=False):
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/rllib/optimizers/multi_gpu.py]: I actually think we don't need this; we should just remove the case-work on L67 and always call `self._init_extra_ops()`

File: python/ray/rllib/ppo/ppo.py [hunk 20]
```diff
@@ -126,133 +129,32 @@ def _init(self):
...
         res = res._replace(info=info)
         return res
 
+    # TODO: fix for local evaluator
     def _fetch_metrics_from_remote_evaluators(self):
         episode_rewards = []
         episode_lengths = []
         metric_lists = [a.get_completed_rollout_metrics.remote()
                         for a in self.remote_evaluators]
+        # metric_lists.append(self.local_evaluator.get_completed_rollout_metrics)
```
Comment (richardliaw) [inline python/ray/rllib/ppo/ppo.py]: can you remove this?

File: python/ray/rllib/ppo/ppo.py [hunk 21]
```diff
@@ -126,133 +129,32 @@ def _init(self):
...
+        info = self.optimizer.stats()
+        # Updates local evaluator kl for computing loss
+        info["kl_coefficient"] = self.local_evaluator.update_kl(info['kl'])
         FilterManager.synchronize(
             self.local_evaluator.filters, self.remote_evaluators)
         res = self._fetch_metrics_from_remote_evaluators()
         res = res._replace(info=info)
         return res
 
+    # TODO: fix for local evaluator
```
Comment (richardliaw) [inline python/ray/rllib/ppo/ppo.py]: remove

File: python/ray/rllib/ppo/ppo_evaluator.py [hunk 22]
```diff
@@ -43,6 +43,8 @@ def __init__(self, registry, env_creator, config, logdir, is_remote):
         self.logdir = logdir
         self.env = ModelCatalog.get_preprocessor_as_wrapper(
             registry, env_creator(config["env_config"]), config["model"])
+
+        # TODO(rliaw): Evaluator should not be the model!!
```
Comment (richardliaw) [inline python/ray/rllib/ppo/ppo_evaluator.py]: can remove

File: python/ray/rllib/ppo/ppo_evaluator.py [hunk 23]
```diff
@@ -166,11 +150,13 @@ def apply_gradients(self, grads):
 
     def save(self):
         filters = self.get_filters(flush_after=True)
-        return pickle.dumps({"filters": filters})
+        return pickle.dumps({"filters": filters,
+                             "kl_coeff": self.config["kl_coeff"]})
```
Comment (richardliaw) [inline python/ray/rllib/ppo/ppo_evaluator.py]: `self.kl_coeff_val`

File: python/ray/rllib/ppo/ppo_evaluator.py [hunk 24]
```diff
@@ -166,11 +150,13 @@ def apply_gradients(self, grads):
...
 
     def restore(self, objs):
         objs = pickle.loads(objs)
         self.sync_filters(objs["filters"])
+        self.kl_coeff = objs["kl_coeff"]
```
Comment (richardliaw) [inline python/ray/rllib/ppo/ppo_evaluator.py]: `self.kl_coeff_val`; `self.kl_coeff` is a tensorflow placeholder

**Review by richardliaw — DISMISSED** [review]

File: python/ray/rllib/optimizers/multi_gpu_impl.py [hunk 25]
```diff
@@ -209,13 +214,19 @@ def _setup_device(self, device, device_input_placeholders):
...
+                    colocate_gradients_with_ops=True)
             return Tower(
                 tf.group(*[batch.initializer
                            for batch in device_input_batches]),
                 device_grads,
                 device_loss_obj)
 
+    def get_shared_weights(self, sess):
+        return {name: v.eval(session=sess)
+                for name, v in self._trainable_vars.items()}
```
Comment (ericl) [inline python/ray/rllib/optimizers/multi_gpu_impl.py]: Why is this necessary?

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 26]
```diff
@@ -42,16 +45,16 @@ def _init(self, sgd_batch_size=128, sgd_stepsize=5e-5, num_sgd_iter=10):
...
+        self._cur_iter_stats = defaultdict(list)
 
         print("LocalMultiGPUOptimizer devices", self.devices)
         print("LocalMultiGPUOptimizer batch size", self.batch_size)
 
         # List of (feature name, feature placeholder) tuples
         self.loss_inputs = self.local_evaluator.tf_loss_inputs()
-
         # per-GPU graph copies created below must share vars with the policy
-        tf.get_variable_scope().reuse_variables()
```
Comment (ericl) [inline python/ray/rllib/optimizers/multi_gpu.py]: Why remove this?

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 27]
```diff
@@ -60,51 +63,90 @@ def _init(self, sgd_batch_size=128, sgd_stepsize=5e-5, num_sgd_iter=10):
...
+        self.local_evaluator.set_weights(weights)
+        if self.remote_evaluators:
+            weight_id = ray.put(weights)
+            for e in self.remote_evaluators:
+                e.set_weights.remote(weight_id)
+
+    def step(self, postprocess_samples=None):
+        """Performs mini-batch SGD.
+
+        This does not update the weights of the local_evaluator.
```
Comment (ericl) [inline python/ray/rllib/optimizers/multi_gpu.py]: Doesn't sync weights update the weights of the local evaluator?

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 28]
```diff
@@ -60,51 +63,90 @@ def _init(self, sgd_batch_size=128, sgd_stepsize=5e-5, num_sgd_iter=10):
...
+        self._init_extra_ops()
+
         self.sess = self.local_evaluator.sess
         self.sess.run(tf.global_variables_initializer())
-
-    def step(self):
+        self._sync_weights()
+
+    def _init_extra_ops(self):
+        self.extra_ops_dict = OrderedDict(self.local_evaluator.tf_extra_ops(
```
Comment (ericl) [inline python/ray/rllib/optimizers/multi_gpu.py]: inline this into `__init__`

File: python/ray/rllib/optimizers/policy_evaluator.py [hunk 29]
```diff
@@ -145,3 +145,6 @@ def build_tf_loss(self, input_placeholders):
         """
 
         raise NotImplementedError
+
+    def tf_extra_ops(self):
+        return {}
```
Comment (ericl) [inline python/ray/rllib/optimizers/policy_evaluator.py]: What interface is this supposed to provide?

File: python/ray/rllib/utils/sampler.py [hunk 30]
```diff
@@ -132,6 +134,34 @@ def get_metrics(self):
         return completed
 
 
+def collect_samples(agents, min_samples):
```
Comment (ericl) [inline python/ray/rllib/utils/sampler.py]: Let's just keep this where it is in rollout.py, we're going to remove it next anyways.

File: python/ray/rllib/utils/common.py [hunk 31]
```diff
@@ -0,0 +1,9 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+
+def standardized(value):
+    # Divide by the maximum of value.std() and 1e-4
+    # to guard against the case where all values are equal
+    return (value - value.mean()) / max(1e-4, value.std())
```
Comment (ericl) [inline python/ray/rllib/utils/common.py]: Inline this in ppo.py

File: python/ray/rllib/optimizers/multi_gpu.py (hunk 26 above)
Comment (richardliaw) [reply python/ray/rllib/optimizers/multi_gpu.py]: the main issue was just that `TOWER_SCOPE` in `multi_gpu_impl` would introduce a different scope, where variables then could not be reused from the local evaluator. 

Basically, there are three places where the policy graph is built: 
 - the local evaluator
 - the `shared_loss` in the `par_opt`
 - the `device_loss`es in the `par_opt`

`tf.get_variable_scope().reuse_variables()` gets the scope where the local_evaluator copy resides, but the `TOWER_SCOPE_NAME` gates off the `par_opt` copies. 

On a separate branch right now, I tried making the local evaluator and `shared_loss` the same copy, but it doesn't work yet and I think I need an extra sync between the `device_losses` graph and the local evaluator graph...

**Comment by skyler0712** [comment]
Are you guys ever going to get done working on my account

Skyler Bays

File: python/ray/rllib/utils/seed.py [hunk 32]
```diff
@@ -0,0 +1,13 @@
+from __future__ import absolute_import
```
Comment (richardliaw) [inline python/ray/rllib/utils/seed.py]: we should support seeding (this utility was incredibly useful during debugging). If so we also need to support environment seeding, which is not covered in this utility.

File: python/ray/rllib/optimizers/multi_gpu.py [hunk 33]
```diff
@@ -72,34 +83,46 @@ def step(self):
...
             if self.remote_evaluators:
-                samples = SampleBatch.concat_samples(
-                    ray.get(
-                        [e.sample.remote() for e in self.remote_evaluators]))
+                # samples = SampleBatch.concat_samples(
+                #     ray.get(
+                #         [e.sample.remote() for e in self.remote_evaluators]))
+                from ray.rllib.ppo.rollout import collect_samples
+                samples = collect_samples(self.remote_evaluators,
+                                          self.timesteps_per_batch)
```
Comment (richardliaw) [inline python/ray/rllib/optimizers/multi_gpu.py]: there's probably a better treatment for this?

File: python/ray/rllib/ppo/ppo_evaluator.py [hunk 34]
```diff
@@ -132,45 +84,64 @@ def build_loss(obs, vtargets, advs, acts, plog, pvf_preds):
...
+
+    def build_tf_loss(self, input_placeholders):
+        obs, vtargets, advs, acts, plog, pvf_preds = input_placeholders
+        return ProximalPolicyGraph(
+            self.env.observation_space, self.env.action_space,
+            obs, vtargets, advs, acts, plog, pvf_preds, self.logit_dim,
+            self.kl_coeff, self.distribution_class, self.config,
+            self.sess, self.registry)
+
+    def init_extra_ops(self, device_losses):
```
Comment (richardliaw) [inline python/ray/rllib/ppo/ppo_evaluator.py]: This was odd because of the chain of dependencies:

1. Local evaluator creates a model
2. MultiGPU creates Variable replicas (which are just refs to the local model)
3. These ops are created, which use nodes from (2)

**Comment by richardliaw** [comment]
@ericl this is ready for readthrough

**Review by ericl — APPROVED** [review]
Looks fine. I'm not too worried about the ugly bits since we'll need to port this to `CommonPolicyEvaluator` next anyways....
